controllers/purge_controller.go

```diff
@@ -0,0 +1,209 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/kyma-project/lifecycle-manager/pkg/log"
+	"github.com/kyma-project/lifecycle-manager/pkg/status"
+	"k8s.io/client-go/rest"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1beta1"
+	"github.com/kyma-project/lifecycle-manager/pkg/adapter"
+	"github.com/kyma-project/lifecycle-manager/pkg/remote"
+	"github.com/kyma-project/lifecycle-manager/pkg/signature"
+	"github.com/kyma-project/lifecycle-manager/pkg/watcher"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlLog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type RemoteClientResolver func(context.Context, client.ObjectKey) (client.Client, error)
+
+// PurgeReconciler reconciles a Kyma object.
+type PurgeReconciler struct {
+	client.Client
+	record.EventRecorder
+	RequeueIntervals
+	signature.VerificationSettings
+	SKRWebhookManager     watcher.SKRWebhookManager
+	KcpRestConfig         *rest.Config
+	RemoteClientCache     *remote.ClientCache
+	ResolveRemoteClient   RemoteClientResolver
+	PurgeFinalizerTimeout time.Duration
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *PurgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrlLog.FromContext(ctx)
+	logger.V(log.InfoLevel).Info("Purge Reconciliation started")
+
+	ctx = adapter.ContextWithRecorder(ctx, r.EventRecorder)
+
+	// check if kyma resource exists
+	kyma := &v1beta1.Kyma{}
+	if err := r.Get(ctx, req.NamespacedName, kyma); err != nil {
+		if apierrors.IsNotFound(err) {
+			remoteClient, err := r.ResolveRemoteClient(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name})
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			if done, msg := performCleanup(ctx, remoteClient, logger); done {
+				logger.Info(fmt.Sprintf("%s", msg))
+			} else {
+				logger.Info("Purging was not successful!")
+				logger.Info(fmt.Sprintf("%s", msg))
+				//TODO: Should we requeue? If the error is permanent, we may end up in the endless loop...
+			}
+
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	// condition to check if deletionTimestamp is set, retry until it gets fully deleted
+	if !kyma.DeletionTimestamp.IsZero() || kyma.Status.State == v1beta1.StateDeleting {
```

Comment:
We should only perform clean-up if `!kyma.DeletionTimestamp.IsZero()` and don't care about `v1beta1.StateDeleting`!